rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,389 @@
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: carto.run/v1alpha1
+      kind: Runnable
+      metadata:
+         name: $(workload.metadata.name)$-linter
+      spec:
+         runTemplateRef:
+            name: md-linting-pipelinerun
+         inputs:
+            repository: $(workload.spec.source.git.url)$
+            revision: $(workload.spec.source.git.revision)$
+         serviceAccountName: pipeline-run-management-sa
+   urlPath: .status.outputs.url
+   revisionPath: .status.outputs.revision
+```
+
+## Proposed example
+Under this proposal, users would only specify one object with a new immutable
+flag.
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate    # <== Now a source template
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         name: $(workload.metadata.name)$-linter      # <== Refers to workload
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$      # <== Refers to workload
+            - name: revision
+              value: $(workload.spec.source.git.revision)$ # <== Refers to workload
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   lifecycle: immutable                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the immutable flag, rather than
+creating/updating the templated object it will create/update a Runnable and
+ensure that a matching ClusterRunTemplate exists.
+
+What is meant by "ensure that a matching ClusterRunTemplate exists"? For any
+given supply chain template (a cluster scoped object) there will only need
+to be one ClusterRunTemplate. All of the N Runnables created by N workloads
+will reference this one ClusterRunTemplate. Only if the supply chain
+template is updated will this ClusterRunTemplate object need updating.
```

Comment:
An advantage of this RFC (so far) is that it does not entail any change to the Runnable controller. To be clear, I am quite willing to change the Runnable controller. But the design of Runnable was contentious in the first place, and I'm wary of picking fights that can be avoided.

That said, let's talk about building the right thing, rather than the institutionally easy thing!

## Must we require a ClusterRunTemplate (CRT)
What would a Runnable look like with an embedded template? Let's transform [this example](https://github.com/vmware-tanzu/cartographer/blob/rfc-simplify-runnable-usage/rfc/rfc-0000-simplify-runnable-use.md#current-example)

We would drop the `templateRef` field from the Runnable and add a `template` field. As you note, we could get rid of the `inputs` field from the Runnable. We would need to add the CRT's `outputs` field to the Runnable. The new object would look like this:
```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterSourceTemplate
metadata:
  name: source-linter
spec:
  template:
    apiVersion: carto.run/v1alpha1
    kind: Runnable
    metadata:
      name: $(workload.metadata.name)$-linter
    spec:
      serviceAccountName: pipeline-run-management-sa
      template:
        apiVersion: tekton.dev/v1beta1
        kind: PipelineRun
        metadata:
          generateName: $(workload.metadata.name)$-pipeline-run-
        spec:
          pipelineRef:
            name: linter-pipeline
          params:
            - name: repository
              value: $(workload.spec.source.git.url)$
            - name: revision
              value: $(workload.spec.source.git.revision)$
          workspaces:
            - name: shared-workspace
              volumeClaimTemplate:
                spec:
                  accessModes:
                    - ReadWriteOnce
                  resources:
                    requests:
                      storage: 256Mi
      outputs:
        url: spec.params[?(@.name=="repository")].value
        revision: spec.params[?(@.name=="revision")].value
  urlPath: .status.outputs.url
  revisionPath: .status.outputs.revision
```

I'll point out for clarity, this is not using the lifecycle flag to create a Runnable behind the scenes. This is an example of the classic method of creating a Runnable, with the proposed removal of the CRT.

As I mentioned, this would be a change to the Runnable controller. Specifically, users would be able to write a Runnable with a templateRef and inputs, and then need to write a CRT. Or users could write a Runnable with a template and outputs.

### Side Question: Could we avoid adding the `outputs` field?
The only alternative that I can see here is to mirror the entirety of the most recently successful stamped object in the status of the Runnable. Perhaps in a .status.mostRecentlySuccessful field (to give users a clue as to what they are reading). In that case the Runnable's `outputs` field could be removed. And the outer `urlPath` would point to `status.mostRecentlySuccessful.spec.params[?(@.name=="repository")].value`. The revisionPath would be similar.

### Result
- Runnable user tutorials becomes more complicated. We have to explain the choice before users. Either 1) use CRT, and Runnable with templateRef and inputs; or 2) use Runnable with template and outputs. Failing that we'd at least have to tell them to ignore the template and output fields they see in the Runnable CRD.
- The logic of the Carto's Runnable controller becomes more complicated. The controller has to handle these two different ways of specifying a Runnable.
- The change to the workload/supplychain controller becomes simpler. Rather than ensuring a single instance of a CRT exists, the workload/supplychain controller can always write Runnables that don't reference CRTs.

Is that tradeoff worth it?

Reply:
I don't know how meaningful this example is, since a user wouldn't create this resource assuming the RFC is otherwise accepted, they'd use the new syntax. A user could create a Runnable independent of the the template resource. In general, I'm not as concerned about the user experience for the Runnable resource as with this change users need to be less aware of it. Logically, the Runnable resource becomes more of an internal implementation detail than a first class.

What I like most about this approach is that a ClusterRunTemplate resource is no longer required. Without this change, the ClusterRunTempaltes would mix user and system resources.

Reply:
I've created a separate RFC for adding a "template" field to the Runnable. See here: https://github.com/vmware-tanzu/cartographer/pull/787